File: src/main/java/com/openlattice/chronicle/ChronicleServer.java
/*
 * Copyright (C) 2018. OpenLattice, Inc.
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program.  If not, see <http://www.gnu.org/licenses/>.
 *
 * You can contact the owner of the copyright at support@openlattice.com
 *
 */

package com.openlattice.chronicle;

import com.dataloom.mappers.ObjectMappers;
import com.fasterxml.jackson.databind.SerializationFeature;
import com.kryptnostic.rhizome.configuration.websockets.BaseRhizomeServer;
import com.kryptnostic.rhizome.core.RhizomeApplicationServer;
import com.kryptnostic.rhizome.hazelcast.serializers.RhizomeUtils.Pods;
import com.kryptnostic.rhizome.pods.hazelcast.RegistryBasedHazelcastInstanceConfigurationPod;
import com.openlattice.auth0.Auth0Pod;
import com.openlattice.aws.AwsS3Pod;
import com.openlattice.chronicle.pods.ChronicleServerSecurityPod;
import com.openlattice.chronicle.pods.ChronicleServerServicesPod;
import com.openlattice.chronicle.pods.ChronicleServerServletsPod;
import com.openlattice.data.serializers.FullQualifiedNameJacksonDeserializer;
import com.openlattice.data.serializers.FullQualifiedNameJacksonSerializer;

public class ChronicleServer extends BaseRhizomeServer {
    public static final Class<?>[] webPods     = new Class<?>[] {
            ChronicleServerServletsPod.class,
            ChronicleServerSecurityPod.class, };
    public static final Class<?>[] rhizomePods = new Class<?>[] {
            RegistryBasedHazelcastInstanceConfigurationPod.class,
            Auth0Pod.class };

    public static final Class<?>[] chronicleServerPods = new Class<?>[] {
            AwsS3Pod.class,
            ChronicleServerServicesPod.class
    };

    static {
        ObjectMappers.foreach( FullQualifiedNameJacksonSerializer::registerWithMapper );
        ObjectMappers.foreach( FullQualifiedNameJacksonDeserializer::registerWithMapper );
        ObjectMappers.foreach( mapper -> mapper.disable( SerializationFeature.WRITE_DATES_AS_TIMESTAMPS ) );
    }

    public ChronicleServer( Class<?>... pods ) {
        super( Pods.concatenate(
                pods,
                webPods,
                rhizomePods,
                RhizomeApplicationServer.DEFAULT_PODS,
                chronicleServerPods ) );
    }

    @Override public void start( String... profiles ) throws Exception {
        super.start( profiles );
    }

    public static void main( String[] args ) throws Exception {
        ChronicleServer chronicleServer = new ChronicleServer();
        chronicleServer.start( args );
    }
}

File: src/main/kotlin/com/openlattice/chronicle/ChronicleServer.kt
package com.openlattice.chronicle

import com.geekbeast.mappers.mappers.ObjectMappers
import com.fasterxml.jackson.databind.ObjectMapper
import com.fasterxml.jackson.databind.SerializationFeature
import com.geekbeast.rhizome.configuration.websockets.BaseRhizomeServer
import com.geekbeast.rhizome.core.RhizomeApplicationServer
import com.geekbeast.rhizome.hazelcast.serializers.RhizomeUtils.Pods
import com.geekbeast.rhizome.pods.hazelcast.RegistryBasedHazelcastInstanceConfigurationPod
import com.geekbeast.auth0.Auth0Pod
import com.geekbeast.aws.AwsS3Pod
import com.openlattice.chronicle.hazelcast.pods.SharedStreamSerializersPod
import com.openlattice.chronicle.mapstores.MapstoresPod
import com.openlattice.chronicle.pods.*
import com.openlattice.chronicle.serializers.FullQualifiedNameJacksonSerializer
import com.openlattice.chronicle.storage.pods.ByteBlobServicePod
import com.openlattice.ioc.providers.LateInitProvidersPod
import com.geekbeast.jdbc.JdbcPod
import com.geekbeast.postgres.PostgresPod
import com.geekbeast.pods.TaskSchedulerPod
import com.openlattice.chronicle.hazelcast.pods.HazelcastQueuePod
import com.openlattice.chronicle.pods.servlet.ChronicleServerSecurityPod
import com.openlattice.chronicle.pods.tables.PostgresDataTablesPod
import com.openlattice.chronicle.pods.tables.PostgresTablesPod
import com.openlattice.chronicle.pods.tables.RedshiftTablesPod

/**
 *
 * @author Matthew Tamayo-Rios &lt;matthew@openlattice.com&gt;
 */
class ChronicleServer(vararg pods: Class<*>) : BaseRhizomeServer(
        *Pods.concatenate(
            pods,
            webPods,
            rhizomePods,
            RhizomeApplicationServer.DEFAULT_PODS,
            chronicleServerPods
        )
) {
    companion object {
        val webPods = arrayOf(
            ChronicleServerServletsPod::class.java,
            ChronicleServerSecurityPod::class.java
        )
        val rhizomePods = arrayOf(
            MapstoresPod::class.java,
            RegistryBasedHazelcastInstanceConfigurationPod::class.java,
            Auth0Pod::class.java
        )
        val chronicleServerPods = arrayOf(
            ChronicleConfigurationPod::class.java,
            AwsS3Pod::class.java,
            JdbcPod::class.java,
            ChronicleServerServicesPod::class.java,
            PostgresPod::class.java,
            PostgresTablesPod::class.java,
            RedshiftTablesPod::class.java,
            PostgresDataTablesPod::class.java,
            TaskSchedulerPod::class.java,
            SharedStreamSerializersPod::class.java,
            ByteBlobServicePod::class.java,
            LateInitProvidersPod::class.java,
            HazelcastQueuePod::class.java,
            ChronicleJobRunnersPod::class.java,
        )

        @Throws(Exception::class)
        @JvmStatic
        fun main(args: Array<String>) {
            val chronicleServer = ChronicleServer()
            chronicleServer.start(*args)
        }

        init {
            ObjectMappers.foreach { mapper: ObjectMapper ->
                FullQualifiedNameJacksonSerializer.registerWithMapper(
                        mapper
                )
            }
            ObjectMappers.foreach { mapper: ObjectMapper ->
                mapper.disable(
                        SerializationFeature.WRITE_DATES_AS_TIMESTAMPS
                )
            }
        }
    }

    @Throws(Exception::class)
    override fun start(vararg profiles: String) {
        super.start(*profiles)
    }
}
